src/sink/http_test.go

```diff
@@ -177,8 +177,8 @@ func defaultPesterClient(t *testing.T) *pester.Client {
 	t.Helper()
 
 	c := pester.New()
-	c.Backoff = pester.LinearBackoff
-	c.MaxRetries = 5
+	c.Backoff = pester.ExponentialBackoff
+	c.MaxRetries = 6
 	c.Timeout = sink.DefaultRequestTimeout
```

Comment:
Is `c.Timeout` enforced as the "global timeout" for all the retries? If so, does it make sense to have both a max amount of retries and a global timeout?

Reply:
`c.Timeout` is an enforced timeout for the connection itself. Meaning the amount of time it is opened a connection.

The global timeout is [here](https://github.com/newrelic/nri-kubernetes/blob/feature/exponential-backoff-for-pester/src/sink/http.go#L16) and is [context who closes the connection](https://github.com/newrelic/nri-kubernetes/blob/feature/exponential-backoff-for-pester/src/sink/http.go#L73).

We could retry until the context is closed instead of having the max retry counted. So it will close when the amount of time is be equal to the exponential backoff plus the theoretically maximum of 15 seconds is more than the context timeout.

The errors are also different:
On context timeout is:
```
DEBU[0068] sending data to httpSink: {"2022-01-05 15:17:22.534716 +0100 CET m=+68.025432250" "Do" "http://localhost:8084/v1/data" "POST" '\x00' '\a' '\x06' "Post \"http://localhost:8084/v1/data\": context deadline exceeded"}
ERRO[0068] publishing integration: performing HTTP request: Post "http://localhost:8084/v1/data": context deadline exceeded
exit status 4
```
While on running without agent is
```
DEBU[0007] sending data to httpSink: {"2022-01-05 15:18:14.276618 +0100 CET m=+7.569047590" "Do" "http://localhost:8084/v1/data" "POST" '\x00' '\x03' '\x02' "Post \"http://localhost:8084/v1/data\": dial tcp 127.0.0.1:8084: connect: connection refused"}
ERRO[0007] publishing integration: performing HTTP request: Post "http://localhost:8084/v1/data": dial tcp 127.0.0.1:8084: connect: connection refused
exit status 4
```

I would say that we almost always will have the `context deadline exceeded` error so we can go directly to retry until we hit the Default Context Timeout. Would you like to add a variable `DefaultCtxTimeout` instead of the `MaxRetries`?

Reply:
I'm still not understanding the difference between `DefaultCtxTimeout` and `c.Timeout`. Which one applies to each connection individually, and which to all attempts globally?

Reply:
`c.Timeout` is the timeout for a single request to happen from the integration to the agent.
`DefaultCtxTimeout` is the timeout for the whole process of connecting once: `DefaultCtxTimeout = backoff + c.Timeout`

Reply:
Neither both of them are used globally for all the attempts because (the context is being close each connection attempt](https://github.com/newrelic/nri-kubernetes/blob/feature/exponential-backoff-for-pester/src/sink/http.go#L74)

Reply:
Ok thanks, it's crystal clear now! It also makes sense why retries is used. Could we add small comments on top of these values explaining this? So we don't get confused when we look at this again :)

Reply:
As far as I remembered was the overall timeout for all requests. 

Notice that the context is closed at the end of each `.Write`. That function use the pester `.Do` that manages all the retries using the same context. This is why there are two:
```
// DefaultCtxTimeout is the default IO timeout for the context of the client.
DefaultCtxTimeout = 15 * time.Second
// DefaultRequestTimeout is the default IO timeout for each request.
DefaultRequestTimeout = 15 * time.Second
```

Moreover, [pester itself](//If the request has been cancelled, skip retries) is request.contex aware stating "//If the request has been cancelled, skip retries" 
If that is not the behaviour than there is a bug since having the two was intended :)